clang/runtime/dpct-rt/include/atomic_ref_ext.hpp.inc

```diff
@@ -0,0 +1,299 @@
+// DPCT_DEPENDENCY_BEGIN
+// AtomicRefExt|atomic_ref_ext
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+// DPCT_LABEL_END
+
+namespace dpct {
+
+// DPCT_LABEL_BEGIN|atomic_ref_ext|dpct
+// DPCT_DEPENDENCY_BEGIN
+// AtomicRefExt|atomic_ref_ext
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+/// Atomic pointer/reference extension to implement standard APIs in std::atomic
+template <typename T,
+          cl::sycl::memory_order DefaultOrder = cl::sycl::memory_order::seq_cst,
+          cl::sycl::memory_scope DefaultScope = cl::sycl::memory_scope::system,
+          cl::sycl::access ::address_space Space =
+              cl::sycl::access::address_space::generic_space>
+class atomic_ref_ext {
+  T __d;
+
+public:
+  /// default memory synchronization order
+  static constexpr cl::sycl::memory_order default_read_order =
+      cl::sycl::detail::memory_order_traits<DefaultOrder>::read_order;
+  static constexpr cl::sycl::memory_order default_write_order =
+      cl::sycl::detail::memory_order_traits<DefaultOrder>::write_order;
+  static constexpr cl::sycl::memory_scope default_scope = DefaultScope;
+  static constexpr cl::sycl::memory_order default_read_modify_write_order =
+      DefaultOrder;
+// DPCT_LABEL_END
+
+// DPCT_LABEL_BEGIN|atomic_ref_ext_1|dpct
+// DPCT_DEPENDENCY_BEGIN
+// AtomicRefExt|atomic_ref_ext
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+  /// no-argument default constructor
+  atomic_ref_ext() = default;
+  /// default constructor
+  atomic_ref_ext(T __d) : __d(__d){};
+// DPCT_LABEL_END
+
+
+// DPCT_LABEL_BEGIN|atomic_ref_ext_store|dpct
+// DPCT_DEPENDENCY_BEGIN
+// AtomicRefExt|atomic_ref_ext
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+  /// atomically replaces the value of the referenced object with a non-atomic argument
+  /// \param [in]  replaces the value of the referenced object
+  /// \param operand The value to replace the pointed value.
+  /// \param memoryOrder The memory ordering used.
+  /// \param memoryScope The memory scope used.
+  void store(T operand, cl::sycl::memory_order memoryOrder = default_write_order,
+             cl::sycl::memory_scope memoryScope = default_scope) const noexcept {
+    cl::sycl::atomic_ref<T, DefaultOrder, DefaultScope, Space> atm(
+        *const_cast<T *>(&__d));
+    atm.store(operand, memoryOrder, memoryScope);
+  }
+// DPCT_LABEL_END
+
+// DPCT_LABEL_BEGIN|atomic_ref_ext_load|dpct
+// DPCT_DEPENDENCY_BEGIN
+// AtomicRefExt|atomic_ref_ext
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+  /// atomically obtains the value of the referenced object
+  /// \param [in, out]  replaces the value of the referenced object
+  /// \param memoryOrder The memory ordering used.
+  /// \param memoryScope The memory scope used.
+  /// \returns The value of the referenced object
+  T load(cl::sycl::memory_order memoryOrder = default_read_order,
+         cl::sycl::memory_scope memoryScope = default_scope) const noexcept {
+    cl::sycl::atomic_ref<T, DefaultOrder, DefaultScope, Space> atm(
+        *const_cast<T *>(&__d));
```

Comment:
How about use 'const_cast<T &>(__d)'?

Reply:
fix it in commits 34e891d3e6db93f6f638490d0805d813e0bf7879